File: javascript/socketd/src/socketd/cluster/ClusterClient.ts
import type {Client} from "../transport/client/Client";
import type { ClientConfig } from "../transport/client/ClientConfig";
import type { ClientSession } from "../transport/client/ClientSession";
import type { Listener } from "../transport/core/Listener";
import type { Session } from "../transport/core/Session";
import type {IoConsumer, IoFunction} from "../transport/core/Typealias";
import {ClusterClientSession} from "./ClusterClientSession";
import {SocketD} from "../SocketD";
import type {ClientConnector} from "../transport/client/ClientConnector";
import {ChannelInternal} from "../transport/core/Channel";

/**
 * 集群客户端
 *
 * @author noear
 */
export class ClusterClient implements Client {
    private _serverUrls: string[];

    private _connectHandler : IoFunction<ClientConnector, Promise<ChannelInternal>>;
    private _heartbeatHandler: IoConsumer<Session>;
    private _configHandler: IoConsumer<ClientConfig>;
    private _listener: Listener;

    constructor(serverUrls: string[] | string) {
        if (serverUrls instanceof Array) {
            this._serverUrls = serverUrls;
        } else {
            this._serverUrls = [serverUrls];
        }
    }

    connectHandler(connectHandler: IoFunction<ClientConnector, Promise<ChannelInternal>>) {
        this._connectHandler = connectHandler;
        return this;
    }

    heartbeatHandler(heartbeatHandler: IoConsumer<Session>): Client {
        this._heartbeatHandler = heartbeatHandler;
        return this;
    }

    /**
     * 配置
     */
    config(configHandler: IoConsumer<ClientConfig>): Client {
        this._configHandler = configHandler;
        return this;
    }

    /**
     * 监听
     */
    listen(listener: Listener): Client {
        this._listener = listener;
        return this;
    }

    async open(): Promise<ClientSession> {
        return this.openDo(false);
    }

    /**
     * 打开
     */
    async openOrThow(): Promise<ClientSession> {
        return this.openDo(true);
    }

    async openDo(isThow: boolean): Promise<ClientSession> {
        const sessionList = new Array<ClientSession>();

        for (const urls of this._serverUrls) {
            for (let url of urls.split(",")) {
                url = url.trim();
                if (!url) {
                    continue;
                }

                const client = SocketD.createClient(url);

                if (this._listener) {
                    client.listen(this._listener);
                }

                if (this._configHandler) {
                    client.config(this._configHandler);
                }

                if (this._connectHandler) {
                    client.connectHandler(this._connectHandler);
                }

                if (this._heartbeatHandler) {
                    client.heartbeatHandler(this._heartbeatHandler);
                }

                if (isThow) {
                    sessionList.push(await client.openOrThow());
                } else {
                    sessionList.push(await client.open());
                }

            }
        }

        return new ClusterClientSession(sessionList);
    }
}

File: python/socketd/socketd/cluster/ClusterClient.py
import asyncio
from typing import List

from socketd import SocketD
from socketd.cluster.ClusterClientSession import ClusterClientSession
from socketd.transport.client.Client import Client, ClientInternal
from socketd.transport.client.ClientConfigHandler import ClientConfigHandler
from socketd.transport.client.ClientConnectHandler import ClientConnectHandler
from socketd.transport.client.ClientSession import ClientSession
from socketd.transport.core import Listener
from socketd.transport.core.Session import Session
from socketd.transport.client.ClientHeartbeatHandler import ClientHeartbeatHandler


class ClusterClient(Client):

    def __init__(self, serverUrls:tuple[str]):
        self.__serverUrls:tuple[str] = serverUrls

        self.__connectHandler: ClientConnectHandler = None
        self.__heartbeatHandler: ClientHeartbeatHandler = None
        self.__configHandler: ClientConfigHandler = None

        self.__listener = None

    def connect_handler(self, connectHandler: ClientConnectHandler)  -> Client:
        self.__connectHandler = connectHandler
        return self

    def heartbeat_handler(self, heartbeatHandler: ClientHeartbeatHandler) -> Client:
        self.__heartbeatHandler = heartbeatHandler
        return self

    def config(self, configHandler: ClientConfigHandler) -> Client:
        self.__configHandler = configHandler
        return self

    def listen(self, listener: Listener) -> Client:
        self.__listener = listener
        return self

    async def open(self) -> ClientSession:
        return await self._open_do(False)

    async def open_or_throw(self) -> ClientSession:
        return await self._open_do(True)

    async def _open_do(self, is_throw):
        sessions: List[Session] = []
        exchangeExecutor = None
        for urls in self.__serverUrls:
            for url in urls.split(","):
                client: ClientInternal = SocketD.create_client(url)

                if self.__listener:
                    client.listen(self.__listener)

                if self.__configHandler:
                    client.config(self.__configHandler)

                if self.__connectHandler:
                    client.connect_handler(self.__connectHandler)

                if self.__heartbeatHandler:
                    client.heartbeat_handler(self.__heartbeatHandler)

                if exchangeExecutor is None:
                    exchangeExecutor = client.get_config().get_exchange_executor()
                else:
                    client.get_config().exchange_executor(exchangeExecutor)

                sesssion = await (client.open_or_throw() if is_throw else client.open())
                sessions.append(sesssion)
        return ClusterClientSession(sessions)


